backend/django/core/forms.py

```diff
@@ -64,45 +79,82 @@ def clean_data_helper(data, supplied_labels):
     except ParserError:
         # If there was an error while parsing then raise invalid file error
         raise ValidationError(
-            "Unable to read file.  " "Please ensure it passes all the requirments"
+            "Unable to read file.  Please ensure it passes all the requirments"
         )
     except UnicodeDecodeError:
         # Some files are not in utf-8, let's just reject those.
         raise ValidationError(
-            "Unable to read the file.  Please "
-            "ensure that the file is encoded in UTF-8."
+            "Unable to read the file.  Please ensure that the file is encoded in UTF-8."
         )
 
     if (len(data.columns) != len(ALLOWED_HEADER)) and len(data.columns) != len(
         ALLOWED_HEADER_ID
     ):
         raise ValidationError(
-            "File has incorrect number of columns.  "
-            "Received {0} but expected {1} or {2}.".format(
+            "File has incorrect number of columns.  Received {0} but expected {1} or {2}.".format(
                 len(data.columns), len(ALLOWED_HEADER), len(ALLOWED_HEADER_ID)
             )
         )
 
-    if (data.columns.tolist() != ALLOWED_HEADER) and (
-        data.columns.tolist() != ALLOWED_HEADER_ID
-    ):
-        raise ValidationError(
-            "File headers are incorrect. "
-            " Received {0} but header must be {1} or {2}.".format(
-                ", ".join(data.columns),
-                ", ".join(ALLOWED_HEADER),
-                ", ".join(ALLOWED_HEADER_ID),
+    for col in ALLOWED_HEADER:
+        if col not in data.columns.tolist():
+            raise ValidationError(
+                "File headers are incorrect. Missing column {0}.".format(col)
+            )
+
+        if col != "Label" and data[col].isnull().sum() == len(data):
+            raise ValidationError("Column {0} is completely empty.".format(col))
+
+    for col in data.columns.tolist():
+        if col not in ALLOWED_HEADER_ID:
+            raise ValidationError(
+                "File headers are incorrect.  Extra column {0}.".format(col)
             )
-        )
+
+    # check the data types of the columns (use pandas library)
+    if metadata["has_created_date"]:
+        # check that the dates are strings
+        if data["created_date"].dtype != np.object:
+            raise ValidationError(
+                "ERROR: created_date column should contain"
+                " date strings like 12/12/2019."
+            )
+        # check that the dates can be parsed.
+        missing_dates = data["created_date"].isnull().sum()
+        data["created_date"] = data["created_date"].dropna().apply(dateparser.parse)
```

Comment:
Is this dropping rows that do not have dates with them?  I dont think that we should make the metadata columns mandatory.

Reply:
I think the goal here was just to not call dateparser on nulls, but I'll check it.

Reply:
Yes so since pandas has it's weird indexing, what happens here is that the non-null dates get passed through the parser and the nulls get left alone. I tested this by printing out the created_date column before and after, and the order of the dates stayed the same. If I remove the dropna() then I get an error for cases where the date is missing.